python/cudf_polars/tests/dsl/test_serialization.py

```diff
@@ -0,0 +1,68 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import pickle
+
+import pytest
+
+from polars.polars import _expr_nodes as pl_expr
+
+from cudf_polars.dsl.expressions.boolean import BooleanFunction
+from cudf_polars.dsl.expressions.datetime import TemporalFunction
+from cudf_polars.dsl.expressions.string import StringFunction
+
+
+@pytest.mark.parametrize(
+    "function", [BooleanFunction, TemporalFunction, StringFunction]
+)
+def test_function_name_serialization_all_values(function):
+    # Test serialization and deserialization for all values of function.Name
+    for name in function.Name:
+        serialized_name = pickle.dumps(name)
+        deserialized_name = pickle.loads(serialized_name)
+        assert deserialized_name is name
+
+
+@pytest.mark.parametrize(
+    "function", [BooleanFunction, TemporalFunction, StringFunction]
+)
+def test_function_name_invalid(function):
+    # Test invalid attribute name
+    with pytest.raises(
+        AttributeError, match="type object 'Name' has no attribute 'InvalidAttribute'"
+    ):
+        assert function.Name.InvalidAttribute is function.Name.InvalidAttribute
+
+
+@pytest.mark.parametrize(
+    "function", [BooleanFunction, TemporalFunction, StringFunction]
+)
+def test_from_polars_all_names(function):
+    # Test that all valid names of polars expressions are correctly converted
+    for name in function.Name:
+        polars_function = getattr(pl_expr, function.__name__)
+        polars_function_attr = getattr(polars_function, name.name)
+        cudf_function = function.Name.from_polars(polars_function_attr)
+        assert cudf_function == name
+
+
+@pytest.mark.parametrize(
+    "function", [BooleanFunction, TemporalFunction, StringFunction]
+)
+def test_from_polars_invalid_attribute(function):
+    # Test converting from invalid attribute name
+    with pytest.raises(ValueError, match=f"{function.__name__} required"):
+        function.Name.from_polars("InvalidAttribute")
+
+
+@pytest.mark.parametrize(
+    "function", [BooleanFunction, TemporalFunction, StringFunction]
+)
+def test_from_polars_invalid_polars_attribute(function):
+    # Test converting from polars function with invalid attribute name
+    with pytest.raises(
+        AttributeError, match="type object 'Name' has no attribute 'InvalidAttribute'"
+    ):
+        function.Name.from_polars(f"{function.__name__}.InvalidAttribute")
```

Comment:
These are both, from the type-system's point of view, statically incorrect, right? Because `from_polars` doesn't technically accept a string.

Reply:
Indeed, but because we cover all of the attributes in the `Name` classes implementations we cannot use any of them as an invalid attribute. This test is thus a proxy to what we do internally which is essentially just checking the strings of names, it's not perfect but it allows us to cover some of those lines of the code without having to hack things like adding an attribute to the loaded library, such as `pl_expr.BooleanFunction.InvalidAttribute = None` or something similar. If you have another less hacky idea I'm up to it as well, or if you prefer I could remove the tests entirely.

Reply:
No, makes sense